### ISSUE TITLE
Set up number of nodes and node type

### ISSUE THREAD

**Issue by radicaled42** [elastic/terraform-provider-ec#695]
Hello everyone, I'm having some problems to set up a ES cluster using terraform.
I'm not using the HCL, but I'm using CDKTF in Typescript.

Basically I'm trying to set up a small "example" cluster that I'm using for testing terraform and `CDKTF`

This is the code:
```
    new EcProvider(this, 'elastic_cloud_provider', {});

    const exampleELK = new Deployment(this, 'elastic_cloud_deployment', {
      name: 'example',
      region: 'eu-west-1',
      version: '8.9.0',
      deploymentTemplateId: 'aws-general-purpose-v5',
      elasticsearch: {
        hot: {
          autoscaling: {
            maxSize: "0g",
            maxSizeResource: ""
          },
          zoneCount: 2,
          size: '4g',
          nodeTypeData: 'true',
          nodeTypeMaster: 'true',
          nodeTypeIngest: 'true',
          nodeTypeMl: 'false'
        },
      },
      kibana: {
        zoneCount: 1,
        size: '1g',
      },
    });
```

I have a couple of question, is it possible to set up a number of nodes?
For example, I would like to set up 10 hot nodes, is it possible to do it?

Also, I'm getting this error with the `nodeTypeXXXXX` configuration:
```
               │ Error: Provider produced inconsistent result after apply
               │
               │ When applying changes to ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment), provider
               │ "provider[\"registry.terraform.io/elastic/ec\"]" produced an unexpected new
               │ value: .elasticsearch.hot.node_type_ml: was cty.StringVal("false"), but now
               │ null.
               │
               │ This is a bug in the provider, which should be reported in the provider's own
               │ issue tracker.
               ╵
               ╷
               │ Error: Provider produced inconsistent result after apply
               │
               │ When applying changes to ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment), provider
               │ "provider[\"registry.terraform.io/elastic/ec\"]" produced an unexpected new
               │ value: .elasticsearch.hot.node_type_ingest: was cty.StringVal("true"), but
               │ now null.
               │
               │ This is a bug in the provider, which should be reported in the provider's own
               │ issue tracker.
               ╵
               ╷
               │ Error: Provider produced inconsistent result after apply
               │
               │ When applying changes to ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment), provider
               │ "provider[\"registry.terraform.io/elastic/ec\"]" produced an unexpected new
               │ value: .elasticsearch.hot.node_type_data: was cty.StringVal("true"), but now
               │ null.
               │
               │ This is a bug in the provider, which should be reported in the provider's own
               │ issue tracker.
```

From the issue [663](https://github.com/elastic/terraform-provider-ec/issues/663#issuecomment-1581394626), I got that 
> This was caused by autoscaling configuration we had to specify: max_size and max_size_resource under autoscaling as null explicitly

But as you can see I set up the max_size and max_size_resource as empty and got the error.

Any ideas I would really appreciate

**Comment by tobio** [comment]
tl;dr remove all the `nodeType*` attributes to fix the error you're seeing. 

It looks like you're getting the same error as https://github.com/elastic/terraform-provider-ec/issues/663. `nodeType*` attributes aren't supported in 8.x stack versions, which only supports node roles. Really, you shouldn't need to specify these though, the provider will use the values from the deployment template. 

> I have a couple of question, is it possible to set up a number of nodes?
> For example, I would like to set up 10 hot nodes, is it possible to do it?

Essentially no. Elastic Cloud only allows specifying the total size of a deployment in a single zone rather than individual node counts/sizes. To have 10 hot nodes per zone, you'd have to scale up to 10x the largest node size (~64G depending on the template). What are you trying to test with a specific node count?

**Comment by radicaled42** [comment]
Thanks for the reply, let me ask a few more questions :) 
I understand that nodeType* is not supported with 8.X stack. But correct me if I'm wrong, but the [node_roles](https://registry.terraform.io/providers/elastic/ec/latest/docs/resources/deployment#node_roles) is a read only variable. 
So we are not able to change it, correct?

For example:
```
      elasticsearch: {
        master: {
          autoscaling: {},
          zoneCount: 1,
          size: '1g'
        },
        hot: {
          autoscaling: {},
          zoneCount: 2,
          size: '4g'
        },
      },
      kibana: {
        zoneCount: 1,
        size: '1g',
      },
    });
```

In the example above, there will be a `master` node with the master role, a `kibana` node and the `hot` node that they will have the roles `ingest` and `data`. 
Is this correct?

@tobio regarding your question:

> Essentially no. Elastic Cloud only allows specifying the total size of a deployment in a single zone rather than individual node counts/sizes. To have 10 hot nodes per zone, you'd have to scale up to 10x the largest node size (~64G depending on the template). What are you trying to test with a specific node count?

With the configuration above, I'm getting this error:

```
               │ Error: failed creating deployment
               │
               │   with ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment),
               │   on cdk.tf.json line 126, in resource.ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment):
               │  126:       }
               │
               │ api error: 1 error occurred:
               │        * cluster.dedicated_master_prohibited: Deployment template [General purpose]
               │ requires at least [6] data nodes before dedicated master can be specified.
               │ Found only [2] data nodes in the deployment (resources.elasticsearch[0])
               │
               │
               |
              ╷
               │ Error: failed creating deployment
               │
               │   with ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment),
               │   on cdk.tf.json line 126, in resource.ec_deployment.elastic_cloud_deployment (elastic_cloud_deployment):
               │  126:       }
               │
               │ set "request_id" to
               │ "x8rl6ft0wxx3sfquv18nuucoycwv1oh8xogfvcssrz0vbjw14vr9nne48upniofr" to
               │ recreate the deployment resources
```

As far as I understand I would need to define the number of `hot` node to be able to set up a `master` node.
I understand I would be able to resize the number of node using `autoscaling`, but I would like to set up a number of node from the start to have.

Thanks again :)

**Comment by tobio** [comment]
> As far as I understand I would need to define the number of hot node to be able to set up a master node.

Yea, this is a bit of a sharp edge with the provider at the moment. Ideally, you shouldn't have to care about master nodes unless you wanted a non-default size node. That is, the provider would just handle it. We have #635 to improve the usage around this.

> but I would like to set up a number of node from the start to have.

So you can create a deployment with more than one hot node per zone. You'd just have something like:

```
hot: {
          autoscaling: {},
          zoneCount: 2,
          size: '174g'
        },
```

It looks like hot nodes on `aws-general-purpose-v5` max out at 58G, so you'd have 3 nodes x 2 zones = 6 nodes total. I understand having to do the maths is annoying though. 

Since you're in CDKTF you could calculate the required size in code. You'd have to get the [deployment template](https://www.elastic.co/guide/en/cloud/current/Deployments_-_Templates.html#get-deployment-template-v2) with `show_instance_configurations=true` which will return the discrete node sizes available for the hot tier then multiple the max size by the number of nodes you're after.

> If I need to migrate a Elasticsearch cluster, I would like to have the same number of nodes at the beginning.

Sizing isn't my area, but the normal advice here would be start off looking for similar available RAM/vCPU rather than pure node count. The RAM/CPU ratio is likely different between on prem and Cloud.
